### ISSUE TITLE
Individual parameters in Modules

### ISSUE THREAD

**Issue by vegapit** [LaurentMazare/tch-rs#72]
Quick question I could not answer in the docs, is it possible to declare some specific parameters to be optimised in a module? For example, in Pytorch, the syntax would look something like:
```
class MyModule ( torch.nn.Module ):
    def __init__(self, x1_0, x2_0):
          self.x1 = torch.nn.Parameter( torch.tensor([x1_0]), requires_grad=True)
          self.x2 = torch.nn.Parameter( torch.tensor([x2_0]), requires_grad=True)
    def forward(self, xs, ys):
          return -( ys - torch.mul(self.x1,xs) - torch.mul(self.x2,xs.exp()) ).pow(2.0).sum()
```
I would then optimize the forward function as a loss and derive x1 and x2

**Comment by LaurentMazare** [comment]
You can certainly do so! An example of a module that currently does this would be the linear layer in `src/nn/linear.rs`.
For this use case, here is some example code that is probably a bit simpler. This first defines a module that uses your two variables `x1` and `x2`. Note that this takes a `nn::Path` as argument which allows you to get variables out of a variable store. A module is returned which implements the `forward` method. When called on tensor `xs` it returns `xs * x1 + xs.exp(x2)` (the rest of your python `forward` function is part of the loss which the optimizer will minimize).

Then in `gradient_descent_test` we create a variable store `vs`, a module using `vs`, and an optimizer for `vs` and run gradient descent in a for loop. This tries to minimize the sum of squared errors between the output of the forward pass of your module applied to `xs` and `ys`.

```rust
use tch::{nn, nn::OptimizerConfig, Device, Kind, Reduction, Tensor, kind};
use tch::nn::Module;

fn my_module(p: nn::Path, dim: i64) -> impl nn::Module {
    let x1 = p.zeros("x1", &[dim]);
    let x2 = p.zeros("x2", &[dim]);
    nn::func(move |xs| xs * &x1 + xs.exp() * &x2)
}

#[test]
fn gradient_descent_test() {
    tch::manual_seed(42);
    let vs = nn::VarStore::new(Device::Cpu);
    let my_module = my_module(vs.root(), 7);
    let opt = nn::Sgd::default().build(&vs, 1e-2).unwrap();
    for _idx in 1..50 {
        // Dummy mini-batches...
        let xs = Tensor::zeros(&[7], kind::FLOAT_CPU);
        let ys = Tensor::zeros(&[7], kind::FLOAT_CPU);
        let loss = (my_module.forward(&xs) - ys).pow(2).sum();
        opt.backward_step(&loss);
    }
```

Let me know if this isn't clear or if you need more help.

**Comment by vegapit** [comment]
Very clear, thank you! Understanding what the `VarStore` and `Path` structs actually do is key as they do not have an equivalent in Pytorch I do not think.

**Comment by LaurentMazare** [comment]
Yes in pytorch these are implicit and driven by scope so it kind of all magically works but lots of things get harder because of this, e.g. optimizing on only some of the variables (which would mostly be done by matching on the variable names).
Anyway I'll make sure to include this example in the main readme and as a test.

**Comment by vegapit** [comment]
Well I have started translating some Pytorch scripts to Rust using this library, so if you think it would be useful I can look into writing a tutorial in the process.

**Comment by LaurentMazare** [comment]
Sure, having more tutorials would be very welcome!

**Comment by LaurentMazare** [comment]
Cool, in the meantime I've added this example to the main readme.

Re passing multiple tensors rather than a single one, note that you may want not to use `nn::func` in this case but instead can just write something like the following if you want to use two tensors (something similar is done for the [actor-critic example](https://github.com/LaurentMazare/tch-rs/blob/master/examples/reinforcement-learning/a2c.rs)).
We could also craft some form of `nn::ModuleN` trait for modules using as input an array of tensors but I'm a bit worried of handling the error when the array does not have the proper number of tensors. Having separate `nn::Module2` (for 2 tensors), `nn::Module3` (for 3 tensors), etc would work too but is quite verbose and won't scale if we want multiple output tensors, maybe something like [const generics](https://github.com/rust-lang/rust/issues/44580) would help here.

```rust
fn my_module(p: nn::Path, dim: i64) -> impl Fn(&Tensor, &Tensor) -> Tensor {
    let x1 = p.zeros("x1", &[dim]);
    let x2 = p.zeros("x2", &[dim]);
    move |xs, ys| xs * &x1 + xs.exp() * &x2 + ys
}

#[test]
fn gradient_descent_test() {
    let vs = nn::VarStore::new(Device::Cpu);
    let my_module = my_module(vs.root(), 7);
    let opt = nn::Sgd::default().build(&vs, 1e-2).unwrap();
    for _idx in 1..50 {
        let xs = Tensor::zeros(&[7], kind::FLOAT_CPU);
        let ys = Tensor::zeros(&[7], kind::FLOAT_CPU);
        let loss = (my_module(&xs, &ys) - &ys).pow(2).sum();
        opt.backward_step(&loss);
    }
}
```

**Comment by vegapit** [comment]
I knew I was missing something...

This example of building a module by simply implementing the `Fn` trait is a crucial one. It really shows how flexible the library already is, so I would not be too worried about creating `nn::ModuleN` or any other wrapper in this case. In my opinion, it is best to keep it streamlined while putting efforts on writing docs to clearly present all use cases.

**Comment by LaurentMazare** [comment]
Closing now that this is covered in your tutorial.
